starknet_devnet/blueprints/rpc/structures/payloads.py

```diff
@@ -1006,14 +1006,48 @@ def rpc_map_traces(
     return result
 
 
+def gateway_to_rpc_invocation(invocation_dict: dict) -> Dict[str, Dict]:
+    """
+    Convert function invocation result from gateway to RPC.
+    """
+    if invocation_dict is None:
+        return invocation_dict
+
+    # All unnecessary properties should be removed except this mentioned here.
+    accepted_properties = {
```

Comment:
Is it enough to just take the gateway structures and call them RPC structures? Don't they require some processing?

Reply:
I don't know to be honest.

Reply:
I believe they require some processing, starkware's model for events and messages returned by `calculate_traces_and_fees` is:
```Python
@dataclasses.dataclass(frozen=True)
class OrderedL2ToL1MessageResponse(ValidatedDataclass):
    """
    See datails in OrderedL2ToL1Message's documentation.
    """

    order: int = field(metadata=sequential_id_metadata("L2-to-L1 message order"))
    to_address: str = field(metadata=everest_fields.EthAddressField.metadata("to_address"))
    payload: List[int] = field(metadata=fields.felt_as_hex_list_metadata)
```

```Python
@dataclasses.dataclass(frozen=True)
class OrderedEventResponse(ValidatedDataclass):
    """
    See datails in OrderedEvent's documentation.
    """

    order: int = field(metadata=sequential_id_metadata("Event order"))
    keys: List[int] = field(metadata=fields.felt_as_hex_list_metadata)
    data: List[int] = field(metadata=fields.felt_as_hex_list_metadata)

```

While in the rpc spec, it should be:
```
            "EVENT_CONTENT": {
                "title": "Event content",
                "description": "The content of an event",
                "type": "object",
                "properties": {
                    "keys": {
                        "type": "array",
                        "title": "Keys",
                        "items": {
                            "$ref": "#/components/schemas/FELT"
                        }
                    },
                    "data": {
                        "type": "array",
                        "title": "Data",
                        "items": {
                            "$ref": "#/components/schemas/FELT"
                        }
                    }
                },
                "required": [
                    "keys",
                    "data"
                ]
            },
```

```
            "MSG_TO_L1": {
                "title": "Message to L1",
                "type": "object",
                "properties": {
                    "from_address": {
                        "description": "The address of the L2 contract sending the message",
                        "$ref": "#/components/schemas/FELT"
                    },
                    "to_address": {
                        "title": "To address",
                        "description": "The target L1 address the message is sent to",
                        "$ref": "#/components/schemas/FELT"
                    },
                    "payload": {
                        "description": "The payload of the message",
                        "title": "Payload",
                        "type": "array",
                        "items": {
                            "$ref": "#/components/schemas/FELT"
                        }
                    }
                },
                "required": [
                    "from_address",
                    "to_address",
                    "payload"
                ]
            },
```

So both seem to have an excessive `order` field, and message doesn't have the `from_address` field?

Reply:
Also, there's no `entry_point_selector` in the cairo-lang model:
```Python
class FunctionInvocation(BaseResponseObject, SerializableMarshmallowDataclass):
    """
    A lean version of CallInfo class, containing merely the information relevant for the user.
    """

    # Static info.
    caller_address: int = field(
        metadata=fields.L2AddressField.metadata(field_name="caller_address")
    )
    contract_address: int = field(metadata=fields.contract_address_metadata)
    calldata: List[int] = field(metadata=fields.calldata_as_hex_metadata)
    call_type: Optional[CallType] = field(metadata=nonrequired_optional_metadata)
    class_hash: Optional[int] = field(metadata=fields.OptionalClassHashIntField.metadata())
    selector: Optional[int] = field(metadata=fields.optional_entry_point_selector_metadata)
    entry_point_type: Optional[EntryPointType]

    # Execution info.
    result: List[int] = field(metadata=fields.retdata_as_hex_metadata)
    execution_resources: ExecutionResources
    internal_calls: List["FunctionInvocation"] = field(
        metadata=additional_metadata(
            marshmallow_field=mfields.List(mfields.Nested(lambda: FunctionInvocation.Schema()))
        )
    )
    events: List[OrderedEventResponse]
    messages: List[OrderedL2ToL1MessageResponse]
```

Reply:
@bartekryba thanks for the input :thumbsup:!

Right now we were discussing those mappings with @FabijanC!

Reply:
also `calls` needs to call `gateway_to_rpc_invocation` recursively

Reply:
I think that `entry_point_selector` is an old remnant but I'm waiting for confirmation from starkware.